### ISSUE TITLE
"Random points along line" algorithm execution fails every two calls

### ISSUE THREAD

**Issue by DelazJ** [qgis/QGIS#33156]
I ran the alg multiple times with the same parameters and concluded that it works every two calls and fails as well every two calls. The error log:

> QGIS version: 3.11.0-Master
QGIS code revision: c52a6e711a
Qt version: 5.11.2
GDAL version: 3.1.0dev
GEOS version: 3.8.0-CAPI-1.13.1 
PROJ version: Rel. 7.0.0, March 1st, 2020
Processing algorithm…
Algorithm 'Random points along line' starting…
Input parameters:
{ 'INPUT' : '//serveur//Études/doss/candidatures.gpkg|layername=A_laisons_douces', 'MIN_DISTANCE' : 0, 'OUTPUT' : 'TEMPORARY_OUTPUT', 'POINTS_NUMBER' : 5 }
>
>Traceback (most recent call last):
File "C:/OSGEO4~1/apps/qgis-dev/./python/plugins\processing\algs\qgis\RandomPointsAlongLines.py", line 126, in processAlgorithm
f = next(source.getFeatures(request.setFilterFid(fid).setSubsetOfAttributes([])))
StopIteration
>
>Execution failed after 0.16 seconds
>
>Loading resulting layers
Algorithm 'Random points along line' finished

**Comment by alexbruy** [comment]
Not reproducible on master (ead3b62c1e).

**Comment by pigreco** [comment]
Does not work:
1. I created a temporary linear layer;
2. I started the algorithm

here is the result:
![image](https://user-images.githubusercontent.com/7631137/73373835-88385e80-42b9-11ea-97d1-381c518d4161.png)

**Comment by pigreco** [comment]
Does not work:

I created a temporary linear layer;
I started the algorithm
here is the result:

![image](https://user-images.githubusercontent.com/7631137/73377108-cedc8780-42be-11ea-94b8-83e6e7bef378.png)

**Comment by agiudiceandrea** [comment]
The algorithm always works correctly with shapefile or gml layers but not with e.g. spatialite, geojson, gpkg or memory layers.
It's due to the fact that firsts have the first feature with fid=0, while the latters have the first feature with fid=1 so https://github.com/qgis/QGIS/blob/c89b511f8c72656115df75e4b160d4082a8ee0d6/python/plugins/processing/algs/qgis/RandomPointsAlongLines.py#L125 fails to correctly pick a random feature due to the incorrect lower limit of the fids range (0 to featureCount - 1).

**Comment by nyalldawson** [comment]
Hmm this algorithm has some issues -- we really shouldn't be picking a totally random feature each time, but rather weighting so that longer features are more likely to be selected..

**Comment by agiudiceandrea** [comment]
A very simple patch https://github.com/agiudiceandrea/QGIS/commit/4d8059d68a5ed5ac6b76fe6a57a46072de330364

**Comment by pigreco** [comment]
I just tested, on the master, the @agiudiceandrea patch,
works well with temporary layer

**Comment by agiudiceandrea** [comment]
> we really shouldn't be picking a totally random feature each time, but rather weighting so that longer features are more likely to be selected..

@nyalldawson Currently the algorithm chooses a random line feature and then a random segment of that line and then a random subsegment (both second and third steps could be avoided using QgsGeometry.interpolate directly on the chosen line feature, after #34140) in order to generate a random point (with the minimum distance constraint): i.e. an equal probability mode. A weighted probability mode could be added as a new option: the weight could be the line feature length, or a field value or an expression value.

**Comment by agiudiceandrea** [comment]
> Hmm this algorithm has some issues

In fact, I've found another issue.

https://github.com/qgis/QGIS/blob/19d58bda08c202b29020a3e9c485fac80350fc79/python/plugins/processing/algs/qgis/RandomPointsAlongLines.py#L144-L150

The `if dist > minDistance` check, wrongly prevents the point to be chosen in the first part of a segment (i.e. at a distance from the start vertex of the segment that is less then the minDistance) as it can be seen running the algorithm multiple times on a single line:
![image](https://user-images.githubusercontent.com/16253859/73653408-61977080-4689-11ea-8877-9d50dd998bc0.png)

It works correctly without the `if dist > minDistance` check

![image](https://user-images.githubusercontent.com/16253859/73653132-c7cfc380-4688-11ea-8c8b-37a8437386d0.png)

**Comment by pigreco** [comment]
@agiudiceandrea 
also tried this latest patch, it works as expected.

**Comment by mapos7** [comment]
QGIS version: 3.34.1-Prizren
QGIS code revision: 133927424d
Qt version: 5.15.3
Python version: 3.9.5
GDAL version: 3.8.1
GEOS version: 3.12.1-CAPI-1.18.1
PROJ version: Rel. 9.3.1, December 1st, 2023
PDAL version: 2.6.0 (git-version: 3fced5)

![image](https://github.com/qgis/QGIS/assets/116223739/a2b66623-8459-422a-bd28-dbbb095bfb2d)

ZeroDivisionError: float division by zero no matter if shapefile or temporaly layer. Is this still the issue or wrong parameterisation?
